### DISCUSSION TITLE
app.publish exclude sender in redis sub

### DISCUSSION THREAD

**Discussion by ghost** [uNetworking/uWebSockets.js#732] [Q&A] [answered]
Hello all,

awesome application, thank you for your work. I have a problem that looks like i cant figure it out by myself, so i want ask here for help.

I want to build up my Server code with scaling in mind from the ground up, so i don't have the hassle later. For this, i used the famous Redis pub/sub mechanics. I want to send a message to all clients, except to the client where the message comes from, because he has the data already locally. Here a little bit of code to demonstrate my problem (stripped down to the problem):

_( I have read a few other discussions about this, but could not find a answer to this problem )_

```js
//Redis Subscribe / Publish 
const subscriber = redisClient.duplicate()
const publisher = redisClient.duplicate()

//Redis Subscription for the messages, want to send to all in topic but exclude sender
//Must be used outside of APP context
void (async () => {
	try {
		await subscriber.subscribe('message', (data) => {
			data = JSON.parse(data)
                        ws.publish(data.topic, data.message) // -> this does what i need but i have no access to ws here
			app.publish(data.topic, data.message) // -> here i can not exclude the original sender as far as i know
		})
	} catch (e) {
		console.error(e)
	}
})()

APP.ws("/*", {
	idleTimeout: 16,
  	compression: APP.SHARED_COMPRESSOR,

        open: async (ws) => {
            //join channel
            ws.subscribe("hello")
        },
        message: async (ws, message, isBinary) => {
            //Publish message
            await publisher.publish("message", JSON.stringify({topic: "hello", message: "world"}))
        }
})
```

The functionality that i need has ws.publish, but i cannot access it at the location where i need it, or is there a way? Did i think wrong? Can someone point me in the right direction and help me in this situation?

I can give more information's if needed.

Thank you very much.

**Comment by e3dio** [comment, marked as answer]
```javascript
const wsMap = new Map();

subscriber.subscribe('message', (data) => {
   data = JSON.parse(data);
   const ws = wsMap.get(data.sender);
   (ws || app).publish(data.topic, data.message);
});

app.ws('/', {
   open: (ws) => {
      wsMap.set(ws.id, ws);
   },
   close: (ws) => {
      wsMap.delete(ws.id);
   }
});
```

**Reply by ghost** [reply]
Hello @e3dio,

simple, elegant and it works. I have no idea why i could not figure it out by myself....

Thank you very much.
